pkg/controller/nodes/attr_path_resolver.go

```diff
@@ -0,0 +1,149 @@
+package nodes
+
+import (
+	"context"
+
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/core"
+	"github.com/flyteorg/flytepropeller/pkg/controller/nodes/errors"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+// resolveAttrPathInPromise resolves the literal with attribute path
+// If the promise is chained with attributes (e.g. promise.a["b"][0]), then we need to resolve the promise
+func resolveAttrPathInPromise(ctx context.Context, nodeID string, literal *core.Literal, bindAttrPath []*core.PromiseAtrribute) (*core.Literal, error) {
+	var currVal *core.Literal = literal
+	var tmpVal *core.Literal
+	var err error
+	var exist bool
+	count := 0
+
+	for _, attr := range bindAttrPath {
+		switch currVal.GetValue().(type) {
+		case *core.Literal_Map:
+			tmpVal, exist = currVal.GetMap().GetLiterals()[attr.GetStringValue()]
+			if exist == false {
+				return nil, errors.Errorf(errors.PromiseAttributeResolveError, nodeID, "key [%v] does not exist in literal %v", attr.GetStringValue(), currVal.GetMap().GetLiterals())
+			}
+			currVal = tmpVal
+			count += 1
+		case *core.Literal_Collection:
+			if int(attr.GetIntValue()) >= len(currVal.GetCollection().GetLiterals()) {
+				return nil, errors.Errorf(errors.PromiseAttributeResolveError, nodeID, "index [%v] is out of range of %v", attr.GetIntValue(), currVal.GetCollection().GetLiterals())
+			}
+			currVal = currVal.GetCollection().GetLiterals()[attr.GetIntValue()]
+			count += 1
+		// scalar is always the leaf, so we can break here
+		case *core.Literal_Scalar:
+			break
+		}
+	}
+
+	// resolve dataclass
+	if currVal.GetScalar() != nil && currVal.GetScalar().GetGeneric() != nil {
+		st := currVal.GetScalar().GetGeneric()
+		// start from index "count"
+		currVal, err = resolveAttrPathInPbStruct(ctx, nodeID, st, bindAttrPath[count:])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return currVal, nil
+}
+
+func resolveAttrPathInPbStruct(ctx context.Context, nodeID string, st *structpb.Struct, bindAttrPath []*core.PromiseAtrribute) (*core.Literal, error) {
+
+	var currVal interface{}
+	var tmpVal interface{}
+	var exist bool
+
+	currVal = st.AsMap()
+
+	// Turn the current value to a map so it can be resolved more easily
+	for _, attr := range bindAttrPath {
+		switch currVal.(type) {
```

Comment:
should we use `reflect.TypeOf` instead? defer to @hamersaw

Reply:
I don't feel strongly either way.